io/io/src/TBufferMerger.cxx

```diff
@@ -90,38 +94,28 @@ void TBufferMerger::SetAutoSave(size_t size)
 
 void TBufferMerger::Merge()
 {
-   fBuffered = 0;
-   fMerger.PartialMerge();
-   fMerger.Reset();
+   {
+      std::lock_guard<std::mutex> m(fMergeMutex);
+      {
+         std::lock_guard<std::mutex> q(fQueueMutex);
+
+         while (!fQueue.empty()) {
+            std::unique_ptr<TBufferFile> buffer{fQueue.front()};
+            fMerger.AddAdoptFile(
```

Comment:
Is the code below guaranteed to **not** take the ROOT global lock (or anything other locks)?  If it can take a lock then there is a risk of dead lock (if somebody calls Merge in another thread while already holding that same lock)?

What about doing this in two stage:

```
decltype<fQueue> localQueue;
{
      std::lock_guard<std::mutex> q(fQueueMutex);
      swap(localQueue,fQueue);
}
while (localQueue.empty()) {
   std::unique_ptr<TBufferFile> buffer{fQueue.front()};
    fMerger.AddAdoptFile(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "READ"));
   localQueue.pop();
}
```

Either way, we the new code the fMergeMutex seems redundant.

Reply:
> Is the code below guaranteed to not take the ROOT global lock (or anything other locks)? Either way, we the new code the fMergeMutex seems redundant.

The `fMergeMutex` is not redundant, it's here to prevent access to `fMerger` from multiple threads/tasks simultaneously, which would lead to a crash. Creating a `TMemFile` does involve momentarily holding `gROOTMutex`, but only to add/remove to list of cleanups. As for calling `Merge()`, it can only happen from a `TBufferMergerFile::Write()`, which triggers a push, or as a side effect of the callback function, which may itself trigger another write and push, so we always enter from the same point into `Merge()`. That's why I release the `fMergeMutex` before calling the callback so that when the callback is entered, no locks are being held.

Reply:
> The fMergeMutex is not redundant, it's here to prevent access to fMerger

Fair enough I missed that.

> Creating a TMemFile does involve momentarily holding gROOTMutex
> ...  it can only happen from a TBufferMergerFile::Write() ...

So if for some reasons one of two callers to TBufferMergeFile::Write hold the gROOTMutex, then there is a potential deadlock, isn't it?

So I still think the following would be an improvement
```
decltype<fQueue> localQueue;
{
   std::lock_guard<std::mutex> q(fQueueMutex);
   swap(localQueue,fQueue);
}
std::queue<TMemFile *> newfiles; 
while (localQueue.empty()) {
   std::unique_ptr<TBufferFile> buffer{fQueue.front()};
   newfiles.push_back(new TMemFile(fMerger.GetOutputFileName(), buffer->Buffer(), buffer->BufferSize(), "READ")));
   localQueue.pop();
}
{ 
   std::lock_guard<std::mutex> m(fMergeMutex); 
   while (newfiles.empty()) {
       fMerger.AddAdoptFile(newfiles.front());
      newfiles.pop();
   }
}
```
which assumes AddAdoptFile does not take the ROOT write lock, if it does then fMergeMutex needs to be replaced by taking the ROOT write mutex.

Cheers,
Philippe.